Begin adding support for the `prefers-reduced-motion` setting, Add a11y notes to the animation docs.

kjellr opened the pull request:
As noted by @ryelle in https://github.com/WordPress/gutenberg/issues/8029#issuecomment-466038569, the [`prefers-reduced-motion` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) is a way for websites to respect an OS-level reduce motion setting. 

In this PR, I've added a quick `@mixin` that we can apply as needed to opt-in to this setting. I've also begun adding it to a couple of our animations to start: 

- The menu appear animation added in #13617.
- The modal window fade/move animation added in #9900.

Unfortunately, this can't be added to all animations (yet). For example, the [menu slide-in animation](https://github.com/WordPress/gutenberg/blob/9dc1a33f2b175b639f9678aa94604bfb29b3838a/packages/components/src/animate/style.scss#L30-L43) moves the menu from its off-screen position into its intended position on screen. As written right now, if we remove the animation, the menu will never be moved on screen. 

It also can't be used in cases where the visibility/opacity of an item is only set by the animation. For instance, I tried applying it to [`edit-post__fade-in-animation`](https://github.com/WordPress/gutenberg/blob/88a69507e0bba453fd0f239b4be1567c94953396/assets/stylesheets/_animations.scss#L1-L3), which is used in a number of places to fade  in content. This worked in some cases, but a few elements (like the block mover icons) didn't appear at all when Reduce Motion was enabled. 

Both of those examples are totally solvable, but I wanted to keep this PR simple for now. 🙂

This PR also adds @joedolson and @ryelle's accessibility recommendations from #8029 to the animation design documentation for greater visibility. 

---

**Testing:** 

1. On a Mac, Open `System Preferences > Accessibility`. 
2. Select "Display" from the left-hand column.
3. Check the box labeled  "Reduce motion"
4. Open Gutenberg using Safari or Firefox (These are the only browsers that support this media query at the moment).
5. Make sure that the menus appear with no animation.
5. Make sure that modal windows appear with no animation.

kjellr committed:
Add a mixin for the reduce motion media query.

kjellr committed:
Add reduce motion media query to the appear animation.

kjellr committed:
Add reduce motion media query to the modal fade animation.

kjellr committed:
Add reduce motion media query to the edit post fade-in animation.

kjellr committed:
Revert "Add reduce motion media query to the edit post fade-in animation."

This reverts commit 4503f04b9480ff899b10bd6863e2326b2ac6b9c3.

kjellr committed:
Add a11y considerations to animation docs.

youknowriad commented on packages/components/src/animate/style.scss:
```diff
@@ -1,6 +1,7 @@
 .components-animate__appear {
 	animation: components-animate__appear-animation 0.1s cubic-bezier(0, 0, 0.2, 1) 0s;
 	animation-fill-mode: forwards;
+	@include reduce-motion;
```
What about the second animation of this file. just a few lines under.

kjellr replied:
I looked at that one too, but I think it needs a more complicated fix. That animation fades in & out indefinitely to denote that something's loading or being updated. If we were to remove the animation, I think we'd need to replace it with some other sort of indicator in its place.

kjellr replied:
(It's totally fixable, but I figured it's probably worth its own PR because it also involves a bit more complicated design problem)

youknowriad replied:
I was thinking about `.components-animate__slide-in` which is not a loading animation AFAIK

kjellr replied:
Oh sorry — I should've looked more closely at which file that was. 😄 

I mentioned that one in the description above. It also needs some deeper thought. Since the `is-from-left` class in there moves the element off canvas, the animation is the only thing that brings it back into view. As it's written currently, if we disable the animation the sidebar will remain hidden forever. 😕

youknowriad replied:
What if you use the technique we use in e2e tests

```
* { animation-duration: 0ms !important; }
```

kjellr replied:
Ah, that's an interesting idea! I think that's just as good as `animation: none` in this case. I'll try swapping that in and see how it goes.

kjellr replied:
Oddly enough, that doesn't seem work reliably for me in Safari. It works great in Firefox, but in Safari, it's preventing the sidebar from appearing at all. 🤔

youknowriad replied:
🤷‍♂️ yeah, our tests only run in Chrome

afercia commented:
Discussed during today's accessibility bug-scrub, definitely appreciating some exploration with `prefers-reduced-motion`. Thanks!

kjellr committed:
Set the animation duration for reduced motion to 1ms.

This essentially stops all animation, but still lets us resolve on the final state.

kjellr committed:
Apply reduced motion to our fade, slide, and fullscreen  animations.

kjellr committed:
Apply reduced motion to the publish panel animation.

kjellr commented:
> Unfortunately, this can't be added to all animations (yet). For example, the menu slide-in animation moves the menu from its off-screen position into its intended position on screen. As written right now, if we remove the animation, the menu will never be moved on screen.

Update: I was able expand this PR to cover this instance, along with many other animations, thanks to a trick @youknowriad [shared above](
https://github.com/WordPress/gutenberg/pull/14021#discussion_r259024216). Instead of setting `animation: none`, I changed the reduce motion setting to `animation-duration: 1ms !important;`. Initially, it was suggested to use `0ms`, but Safari didn't seem to land on the final state when I tried that. `1ms` gets the job done, and results in no discernible animation in my tests.  

This PR now effects the following animations:

- The menu appear animation added in #13617.
- The modal window fade/move animation added in #9900.
- The sliding sidebar animation added in #13635
- The fullscreen animation added in #9910
- Our reusable [fade animation](https://github.com/WordPress/gutenberg/pull/14021/files#diff-436652042e10cc8d81d6d2ee2b2fd4a7R8)
- The publish panel [slide animation](https://github.com/WordPress/gutenberg/pull/14021/files#diff-ec5f5c91d0b3295f448977820605e4e4R181)

I'd love another review, as well as thoughts on the documentation update included in this PR as well. Thanks!

ryelle commented on docs/designers-developers/designers/animation.md:
```diff
@@ -28,6 +28,13 @@ In creating consistent animations, we have to establish physical rules for how e
 
+## Accessibility Considerations
+
+- Animations should be subtle. Be cognizent of users with [motion triggering vestibular disorders](https://www.ncbi.nlm.nih.gov/pubmed/29017000).
```
This feels like awkward phrasing, what about
> Animations should be subtle. Be cognizant of users with [vestibular disorders triggered by motion.](https://www.ncbi.nlm.nih.gov/pubmed/29017000)

kjellr replied:
That's much better, thanks! 👍

ryelle commented on docs/designers-developers/designers/animation.md:
```diff
@@ -28,6 +28,13 @@ In creating consistent animations, we have to establish physical rules for how e
+- Animations should be subtle. Be cognizent of users with [motion triggering vestibular disorders](https://www.ncbi.nlm.nih.gov/pubmed/29017000).
+- Don't animate elements that are currently reporting content to adaptive technology (e.g., an `aria-live` region that's receiving updates). This can cause confusion wherein the technology tries to parse a region that's actively changing.
+- Avoid animations that aren't directly triggered by user behaviors.
+- Whenever possible, ensure that animations respect the OS-level "Reduce Motion" settings. This can be done by utilizing the [`prefers-reduce-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion) media query. 
```
Is this documentation for working on gutenberg, or just best practices in general? If the former, you could mention the `reduce-motion` mixin?

kjellr replied:
Good call! Added in `896410e`.

ryelle reviewed (approved):
Checked the animations out in Mac/Safari & it's working great 🙂  Left some notes on the docs, that looks good too– anything more might be too basic and not necessary, like [don't use a rapid flashing animation,](https://www.w3.org/WAI/WCAG21/Understanding/three-flashes-or-below-threshold.html) and don't loop animations (which is covered by the speed guidelines).

kjellr committed:
Clean up vestibular disorder phrasing in documentation.

Props @ryelle

kjellr committed:
Mention the @reduce-motion mixin in the docs.

m-e-h commented on assets/stylesheets/_mixins.scss:
```diff
@@ -333,3 +333,13 @@
+
+@mixin reduce-motion {
+	@media (prefers-reduced-motion: reduce) {
+		animation-duration: 1ms !important;
```
I know I'm late to this merged tickets party but thought I'd mention the lack of `transition-duration: 1ms !important;` in this mixin.

Not sure if a `transition` is even used anywhere but it would make the mixin a little more versatile.

kjellr replied:
Good call — we could try that too. I wonder if it'd make sense to create two different mixins — one for `transition`, and one for `animation`. Just so that we aren't including of unused `!important` rules when we don't need them. I may be overthinking it though. 😄

In any case, seems worth exploring! Feel free to open up a PR.
